scripts/toolchain/dbtenv.sh

```diff
@@ -0,0 +1,343 @@
+    mkdir -p "$DBT_TOOLCHAIN_PATH/toolchain" || return 1;
+    mv "$DBT_TOOLCHAIN_PATH/toolchain/$TOOLCHAIN_DIR" "$TOOLCHAIN_ARCH_DIR" || return 1;
+    echo "done";
+    return 0;
+}
+
+dbtenv_cleanup()
+{
+    printf "Cleaning up..";
+    if [ -n "${DBT_TOOLCHAIN_PATH:-""}" ]; then
+        rm -rf "${DBT_TOOLCHAIN_PATH:?}/toolchain/"*.tar.gz;
+        rm -rf "${DBT_TOOLCHAIN_PATH:?}/toolchain/"*.part;
+    fi
+    echo "done";
+    trap - 2;
+    return 0;
+}
+
+dbtenv_curl_wget_check()
+{
+    printf "Checking curl..";
+    if ! curl --version > /dev/null 2>&1; then
+        echo "no";
+        printf "Checking wget..";
+        if ! wget --version > /dev/null 2>&1; then
+            echo "no";
+            echo "No curl or wget found in your PATH";
+            echo "Please provide it or download this file:";
+            echo;
+            echo "$TOOLCHAIN_URL";
+            echo;
+            echo "And place in $DBT_TOOLCHAIN_PATH/toolchain/ dir manually";
+            return 1;
+        fi
+        echo "yes"
+        DBT_DOWNLOADER="dbtenv_wget";
+        return 0;
+    fi
+    echo "yes"
+    DBT_DOWNLOADER="dbtenv_curl";
+    return 0;
+}
+
+dbtenv_check_download_toolchain()
+{
+    if [ ! -d "$TOOLCHAIN_ARCH_DIR" ]; then
+        dbtenv_download_toolchain || return 1;
+    elif [ ! -f "$TOOLCHAIN_ARCH_DIR/VERSION" ]; then
+        dbtenv_download_toolchain || return 1;
+    elif [ "$(cat "$TOOLCHAIN_ARCH_DIR/VERSION")" != "$DBT_TOOLCHAIN_VERSION" ]; then
+        echo "DBT: starting toolchain upgrade process.."
+        dbtenv_download_toolchain || return 1;
+    fi
+    return 0;
+}
+
+dbtenv_download_toolchain()
+{
+    dbtenv_check_tar || return 1;
+    TOOLCHAIN_TAR="$(basename "$TOOLCHAIN_URL")";
+    TOOLCHAIN_DIR="$(echo "$TOOLCHAIN_TAR" | sed "s/-$DBT_TOOLCHAIN_VERSION.tar.gz//g")";
+    trap dbtenv_cleanup 2;  # trap will be restored in dbtenv_cleanup
+    if ! dbtenv_check_downloaded_toolchain; then
+        dbtenv_curl_wget_check || return 1;
+        dbtenv_download_toolchain_tar || return 1;
+    fi
+    dbtenv_remove_old_tooclhain;
+    dbtenv_unpack_toolchain || return 1;
+    dbtenv_cleanup;
+    return 0;
+}
+
+dbtenv_print_version()
+{
+    if [ -n "$DBT_VERBOSE" ]; then
+        echo "DBT: using toolchain version $(cat "$TOOLCHAIN_ARCH_DIR/VERSION")";
+    fi
+}
+
+dbtenv_main()
+{
+    dbtenv_check_sourced || return 1;
+    dbtenv_get_kernel_type || return 1;
+    if [ "$1" = "--restore" ]; then
+        dbtenv_restore_env;
+        return 0;
+    fi
+    dbtenv_check_if_sourced_multiple_times;
+    dbtenv_check_env_vars || return 1;
+    dbtenv_check_download_toolchain || return 1;
+    dbtenv_set_shell_prompt;
+    dbtenv_print_version;
+    PATH="$TOOLCHAIN_ARCH_DIR/python/bin:$PATH";
+    PATH="$TOOLCHAIN_ARCH_DIR/gcc-arm/bin:$PATH";
+    PATH="$TOOLCHAIN_ARCH_DIR/openocd/bin:$PATH";
+    PATH="$TOOLCHAIN_ARCH_DIR/openssl/bin:$PATH";
```

Comment:
It would be nice to have a switch that turns path hacking off, since binary compatibility on linux is an eldritch hell. On nixos it's easier for me to provide these binaries on the system path than it is to get random downloaded executables working.

Reply:
I hear you, and a switch might be a nice addition, but also the Flipper Zero folks put a lot into making their tool work well in these scenarios. If you'd like to see how it runs, feel free to check out their source and give the flipperdevices/flipperzero-firmware "fbt" command a run. =)

Reply:
```
/home/bob_twinkles/Code/fbt/flipperzero-firmware/toolchain/x86_64-linux/python/bin/python3: line 15: /home/bob_twinkles/Code/fbt/flipperzero-firmware/toolchain/x86_64-linux/python/bin/python3.11-bin: cannot execute: required file not found
```
On `nixos` you can't even trust the path to libc (which is what's happening here), so binaries usually need [`patchelf`](https://github.com/NixOS/patchelf) to run. Or better, built from source as a native nix derivation with the right rpath already embedded. If you don't want to support this kind of weirdness I'll happily carry a local patch or spin up a container instead =P

Reply:
Interesting! I think I'm going to aim to support the "big 3" out of the box, but after the build scripts are working I'm happy to see if we can integrate support for nixos and other unusual platforms.